File: MonotonicStack/Medium/739_Daily_Temperatures/Solution.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        res = [0] * n
        st = []

        for i, num in enumerate(temperatures):
            while st and temperatures[st[-1]] < num:
                res[st[-1]] = i - st[-1]
                st.pop()

            st.append(i)

        return res

File: MonotonicStack/Medium/739_Daily_Temperatures/Solution.cpp
#include <vector>
#include <stack>


class Solution {
public:
    std::vector<int> dailyTemperatures(std::vector<int>& vec) {
        int n = vec.size();
        std::stack<int> st;
        std::vector<int> res(n);

        for (int i{}; i < n; ++i) {
            while (!st.empty() && vec[st.top()] < vec[i]) {
                res[st.top()] = i - st.top();
                st.pop();
            }

            st.push(i);
        }

        return res;
    }
};